cli/src/semdep/parsers/mix.py

```diff
@@ -0,0 +1,223 @@
+from pathlib import Path
+from typing import Any
+from typing import List
+from typing import Optional
+from typing import Set
+from typing import Tuple
+
+from semdep.external.parsy import any_char
+from semdep.external.parsy import Parser
+from semdep.external.parsy import regex
+from semdep.external.parsy import string
+from semdep.external.parsy import success
+from semdep.parsers.util import colon
+from semdep.parsers.util import comma
+from semdep.parsers.util import consume_line
+from semdep.parsers.util import DependencyFileToParse
+from semdep.parsers.util import filter_on_marked_lines
+from semdep.parsers.util import lbrace
+from semdep.parsers.util import lbrack
+from semdep.parsers.util import mark_line
+from semdep.parsers.util import new_lines
+from semdep.parsers.util import quoted_str
+from semdep.parsers.util import rbrace
+from semdep.parsers.util import rbrack
+from semdep.parsers.util import safe_parse_lockfile_and_manifest
+from semdep.parsers.util import transitivity
+from semdep.parsers.util import upto
+from semdep.parsers.util import whitespace
+from semgrep.semgrep_interfaces.semgrep_output_v1 import DependencyParserError
+from semgrep.semgrep_interfaces.semgrep_output_v1 import Ecosystem
+from semgrep.semgrep_interfaces.semgrep_output_v1 import FoundDependency
+from semgrep.semgrep_interfaces.semgrep_output_v1 import Mix
+from semgrep.semgrep_interfaces.semgrep_output_v1 import MixLock
+from semgrep.semgrep_interfaces.semgrep_output_v1 import ScaParserName
+
+# :hex,
+atom = string(":") >> upto(",", consume_other=False)
+
+# Lockfile
+scm = atom
+package_block = atom
+
+# "2.7.0"
+version_block = quoted_str
+
+# [:mix], [], [:mix, :rebar3]
+any_array = lbrack >> any_char.until(rbrack, consume_other=True)
+
+# {:plug, "~> 1.14", [hex: :plug, repo: "hexpm", optional: false]}
+inner_dependency_block = (
+    lbrace >> atom >> comma >> quoted_str >> comma >> any_array >> rbrace
+)
+
+# {:plug, "~> 1.14", [hex: :plug, repo: "hexpm", optional: false]}, {:octo_fetch, "~> 0.3", [hex: :octo_fetch, repo: "hexpm", optional: false]}
+many_independency_blocks = inner_dependency_block.sep_by(comma)
+
+
+#   {
+#     :hex,
+#     :testing,
+#     "1.0.1",
+#     "1234bb4db5b32fc0f8aa5c4a2040348b4aa36687100fb8837b850e90cf60e06",
+#     [:mix],
+#     [],
+#     "hexpm",
+#     "98767a5d1c6c3e3d20497b03293be7f83b46f89a6f3987cc1f9262d299f1eaa7"
+#   }
+def package_entry_hex_value_block(package: str) -> Parser:
+    return (
+        whitespace
+        >> lbrace
+        >> atom
+        >> comma
+        >> atom
+        >> comma
+        >> version_block.bind(lambda version: mark_line(success((package, version))))
+        << comma
+        << quoted_str
+        << comma
+        << any_array
+        << comma
+        << lbrack
+        << many_independency_blocks
+        << rbrack
+        << comma
+        << quoted_str
+        << comma
+        << quoted_str
+        << rbrace
+        << comma.optional()
+    )
```

Comment:
Maybe we should have a `sequence` function that takes a list of parsers to run in sequence 😅